File: app/bim_ai/agent_brief_command_protocol.py
"""Deterministic brief → command plan protocol for agent review surfaces (WP-F01)."""

from __future__ import annotations

from typing import Any

from bim_ai.document import Document
from bim_ai.elements import (
    AgentAssumptionElem,
    AgentDeviationElem,
    BcfElem,
    IssueElem,
)


def agent_brief_command_protocol_v1(
    *,
    doc: Document,
    proposed_commands: list[dict[str, Any]],
    validation_violations: list[dict[str, Any]],
) -> dict[str, Any]:
    """Derive a replayable protocol object from model state, command preview, and violations."""

    bcfs = sorted(
        (e for e in doc.elements.values() if isinstance(e, BcfElem)),
        key=lambda x: x.id,
    )
    issues_open = sorted(
        (
            e
            for e in doc.elements.values()
            if isinstance(e, IssueElem) and e.status != "done"
        ),
        key=lambda x: x.id,
    )
    if bcfs:
        b0 = bcfs[0]
        source_brief: dict[str, Any] = {
            "briefKind": "bcf",
            "briefId": b0.id,
            "briefTitle": b0.title,
        }
    elif issues_open:
        i0 = issues_open[0]
        source_brief = {
            "briefKind": "issue",
            "briefId": i0.id,
            "briefTitle": i0.title,
        }
    else:
        source_brief = {"briefKind": None, "briefId": None, "briefTitle": None}

    assumption_ids = sorted(
        e.id for e in doc.elements.values() if isinstance(e, AgentAssumptionElem)
    )
    assumption_id_set = frozenset(assumption_ids)
    deviation_ids = sorted(
        e.id for e in doc.elements.values() if isinstance(e, AgentDeviationElem)
    )
    missing_refs: list[dict[str, str]] = []
    for e in sorted(
        (x for x in doc.elements.values() if isinstance(x, AgentDeviationElem)),
        key=lambda x: x.id,
    ):
        rid = e.related_assumption_id
        if rid is not None and rid.strip() and rid not in assumption_id_set:
            missing_refs.append({"deviationId": e.id, "relatedAssumptionId": rid})
    missing_refs.sort(key=lambda x: (x["deviationId"], x["relatedAssumptionId"]))

    hist: dict[str, int] = {}
    for c in proposed_commands:
        if not isinstance(c, dict):
            key = "?"
        else:
            t = c.get("type")
            key = str(t) if t is not None else "?"
        hist[key] = hist.get(key, 0) + 1
    command_type_histogram = dict(sorted(hist.items()))

    rule_ids: set[str] = set()
    blocker_rules: set[str] = set()
    target_element_ids: set[str] = set()
    for raw in validation_violations:
        if not isinstance(raw, dict):
            continue
        rid = raw.get("ruleId")
        if isinstance(rid, str) and rid:
            rule_ids.add(rid)
            if raw.get("blocking") is True or raw.get("severity") == "error":
                blocker_rules.add(rid)
        eids_raw = raw.get("elementIds")
        if isinstance(eids_raw, list):
            for x in eids_raw:
                if isinstance(x, str) and x:
                    target_element_ids.add(x)

    return {
        "format": "agentBriefCommandProtocol_v1",
        "schemaVersion": 1,
        "sourceBrief": source_brief,
        "assumptionIds": assumption_ids,
        "deviationIds": deviation_ids,
        "missingAssumptionReferences": missing_refs,
        "proposedCommandCount": len(proposed_commands),
        "commandTypeHistogram": command_type_histogram,
        "validationRuleIds": sorted(rule_ids),
        "validationTargetElementIds": sorted(target_element_ids),
        "unresolvedBlockers": sorted(blocker_rules),
    }

File: packages/web/src/workspace/agent/agentBriefCommandProtocol.ts
export type AgentBriefSourceBrief = {
  briefKind: string | null;
  briefId: string | null;
  briefTitle: string | null;
};

export type AgentBriefCommandProtocolV1 = {
  format: 'agentBriefCommandProtocol_v1';
  schemaVersion: number;
  sourceBrief: AgentBriefSourceBrief;
  assumptionIds: string[];
  deviationIds: string[];
  missingAssumptionReferences: { deviationId: string; relatedAssumptionId: string }[];
  proposedCommandCount: number;
  commandTypeHistogram: Record<string, number>;
  validationRuleIds: string[];
  validationTargetElementIds: string[];
  unresolvedBlockers: string[];
};

function isRecord(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null && !Array.isArray(x);
}

/** Parse server payload field into a typed protocol, or null if missing/invalid. */
export function parseAgentBriefCommandProtocolV1(raw: unknown): AgentBriefCommandProtocolV1 | null {
  if (!isRecord(raw)) return null;
  if (raw.format !== 'agentBriefCommandProtocol_v1') return null;
  const schemaVersion = raw.schemaVersion;
  if (typeof schemaVersion !== 'number') return null;

  const sb = raw.sourceBrief;
  if (!isRecord(sb)) return null;
  const briefKind = typeof sb.briefKind === 'string' ? sb.briefKind : null;
  const briefId = typeof sb.briefId === 'string' ? sb.briefId : null;
  const briefTitle = typeof sb.briefTitle === 'string' ? sb.briefTitle : null;

  if (!Array.isArray(raw.assumptionIds) || !raw.assumptionIds.every((x) => typeof x === 'string'))
    return null;
  if (!Array.isArray(raw.deviationIds) || !raw.deviationIds.every((x) => typeof x === 'string'))
    return null;
  const missing = raw.missingAssumptionReferences;
  if (!Array.isArray(missing)) return null;
  for (const row of missing) {
    if (
      !isRecord(row) ||
      typeof row.deviationId !== 'string' ||
      typeof row.relatedAssumptionId !== 'string'
    )
      return null;
  }
  if (typeof raw.proposedCommandCount !== 'number') return null;
  const hist = raw.commandTypeHistogram;
  if (!isRecord(hist)) return null;
  for (const v of Object.values(hist)) {
    if (typeof v !== 'number') return null;
  }
  if (
    !Array.isArray(raw.validationRuleIds) ||
    !raw.validationRuleIds.every((x) => typeof x === 'string')
  )
    return null;
  if (
    !Array.isArray(raw.validationTargetElementIds) ||
    !raw.validationTargetElementIds.every((x) => typeof x === 'string')
  )
    return null;
  if (
    !Array.isArray(raw.unresolvedBlockers) ||
    !raw.unresolvedBlockers.every((x) => typeof x === 'string')
  )
    return null;

  return {
    format: 'agentBriefCommandProtocol_v1',
    schemaVersion,
    sourceBrief: { briefKind, briefId, briefTitle },
    assumptionIds: [...raw.assumptionIds],
    deviationIds: [...raw.deviationIds],
    missingAssumptionReferences: missing.map((row) => {
      const r = row as Record<string, unknown>;
      return {
        deviationId: r.deviationId as string,
        relatedAssumptionId: r.relatedAssumptionId as string,
      };
    }),
    proposedCommandCount: raw.proposedCommandCount,
    commandTypeHistogram: { ...hist } as Record<string, number>,
    validationRuleIds: [...raw.validationRuleIds],
    validationTargetElementIds: [...raw.validationTargetElementIds],
    unresolvedBlockers: [...raw.unresolvedBlockers],
  };
}

/** Deterministic monospace lines for UI + Vitest snapshots. */
export function formatAgentBriefCommandProtocolReadout(
  protocol: AgentBriefCommandProtocolV1 | null,
): string[] {
  if (protocol === null) return ['(no protocol)'];

  const histLines = Object.keys(protocol.commandTypeHistogram)
    .sort()
    .map((k) => `  hist ${k}: ${protocol.commandTypeHistogram[k]}`);

  const miss = protocol.missingAssumptionReferences
    .slice()
    .sort(
      (a, b) =>
        a.deviationId.localeCompare(b.deviationId) ||
        a.relatedAssumptionId.localeCompare(b.relatedAssumptionId),
    )
    .map((m) => `  missingRef ${m.deviationId} -> ${m.relatedAssumptionId}`);

  return [
    `format: ${protocol.format}`,
    `schemaVersion: ${protocol.schemaVersion}`,
    `brief: ${protocol.sourceBrief.briefKind ?? '—'} / ${protocol.sourceBrief.briefId ?? '—'} / ${protocol.sourceBrief.briefTitle ?? '—'}`,
    `assumptionIds: ${protocol.assumptionIds.join(', ') || '—'}`,
    `deviationIds: ${protocol.deviationIds.join(', ') || '—'}`,
    ...(miss.length
      ? ['missingAssumptionReferences:', ...miss]
      : ['missingAssumptionReferences: —']),
    `proposedCommandCount: ${protocol.proposedCommandCount}`,
    ...(histLines.length ? ['commandTypeHistogram:', ...histLines] : ['commandTypeHistogram: —']),
    `validationRuleIds: ${protocol.validationRuleIds.join(', ') || '—'}`,
    `validationTargetElementIds: ${protocol.validationTargetElementIds.join(', ') || '—'}`,
    `unresolvedBlockers: ${protocol.unresolvedBlockers.join(', ') || '—'}`,
  ];
}
